### PR TITLE
UI-8056 - Introduce getMigrateFilters function

### PR DESCRIPTION
**Pull request by eliasinho**
## Description

Introduce a `getMigrateFilters` function, supposed to be generic, that:

- returns a migration function that takes a `callback` argument. This callback function handles the actual migration of the filters, the logic in it depending on which migration is done
- mutates `contentServer` and `counters`.
- handles error cases by mutating `errorReport`

In `bin.ts`, a curried `migrateFilters` function will be created thanks to `getMigrateFilters ` and passed to `migrate_fromVersion_to_toVersion`. The global variables meant to be mutated are also defined at the level of `bin.ts` and are passed to `migrate_fromVersion_to_toVersion`.

In `migrate_fromVersion_to_toVersion`, the filters migration is handled in the callback passed to `migrateFilters`, passed in the arguments. `migrate_fromVersion_to_toVersion` returns a `void`.

### CODE REVIEW

**Review by antoinetissier — COMMENTED** [review]
I think ultimately we should drop the `MigrateXyzCallback` types: it will be used only in `getMigrateXyz` functions, and it adds an unnecessary indirection to just inline with the right type generics imo.

Finally, independently from the above, more arguments need to be passed to the callback (and therefore to the `getMigrateXyz` function in the first place):  `servers` (we could take the opportunity to rename it to `dataModels`) for all content types, and `keysOfWidgetPluginsToRemove` for dashboards.
Let's do that in a separate PR.

File: src/getMigrateDashboards.ts [hunk 1]
```diff
@@ -56,6 +54,7 @@ export const getMigrateDashboards =
         // The dashboard could not be migrated at all.
         counters.dashboards.failed++;
 
+        const filesAncestry = _getFilesAncestry(dashboardsStructure);
```
[the remark this answers is not in the document]
Comment (eliasinho) [reply src/getMigrateDashboards.ts:57]: Yeah, good point, it's more efficient outside the loop, my bad!

**Comment by eliasinho** [comment]
> More arguments need to be passed to the callback (and therefore to the `getMigrateXyz` function in the first place): `dataModels` for all content types, and `keysOfWidgetPluginsToRemove` for widgets.

Okay, I'll add them. Can you explain why we need them?

**Comment by antoinetissier** [comment]
> > More arguments need to be passed to the callback (and therefore to the `getMigrateXyz` function in the first place): `dataModels` for all content types, and `keysOfWidgetPluginsToRemove` for widgets.
> 
> Okay, I'll add them. Can you explain why we need them?

We need the `dataModels` for any operation including a cube lookup, which in particular includes almost any MDX manipulation.
We need the `keysOfWidgetPluginsToRemove` to allow callers to ignore some widgets such as custom ones.

You can see example in the migration from `4.3` to `5.0`.

File: src/getMigrateFilters.ts [hunk 2]
```diff
@@ -25,10 +25,14 @@ export const getMigrateFilters =
       errorReport,
       counters,
       doesReportIncludeStacks,
+      // `dataModels` is not used yet, but needs to be in the function's signature.
+      // eslint-disable-next-line
```
Comment (antoinetissier) [inline src/getMigrateFilters.ts:29]: The point is to pass it to the callback: it wouldn't be useful if we just put it in the signature.

But if you want let's just merge without the `dataModels` for now.
I have it ready on a subsequent branch, so we can add it from there.

**Comment by antoinetissier** [comment]
I pushed a commit to leave adding the `dataModels` out of the PR for now.
Looks good to me like this !

**Pull request merged**